Adding reset password confirmation page

acbeni opened the pull request:
Closes #651

lauraluiz commented on app/controllers/myaccount/SuccessfulPasswordRecoverController.java:
```diff
@@ -0,0 +1,58 @@
+        TemplateControllerComponentsSupplier.class,
+        PageHeaderControllerComponentSupplier.class
+})
+public final class SuccessfulPasswordRecoverController extends SunriseContentController implements WithQueryFlow<Void> {
```
I just realized that you used a different Controller. How is it? I thought you were going to just add a "sucessful" action to the `RecoverPasswordController` and modify the template so that it would show a successful message, optionally hiding the form.

lauraluiz commented on common/app/com/commercetools/sunrise/framework/reverserouters/myaccount/recoverpassword/SimpleRecoverPasswordReverseRouter.java:
```diff
@@ -23,4 +23,8 @@
+
+    String RESET_PASSWORD_SUCCESS = "resetPasswordSuccessCall";
+
+    Call resetPasswordSuccessCall(final String languageTag);
```
The reset/recovery naming is always quite confusing, but here I would call it `requestRecoveryEmailSuccessCall`. Notice that this is the success scenario from the `requestRecoveryEmailProcessCall` (when you submit your email), not from the `resetPasswordProcessCall` (which would be when you change your password).

acbeni replied:
Yeo i agreed with you 👍 
i will change the name accordingly  to be in harmony with the naming scheme

lauraluiz commented on conf/templates/common/additional-html-head.hbs:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+<!-- start common/additional-html-head.hbs -->
+
+<!-- end common/additional-html-head.hbs -->
```
Why did you change these files? (Same for `additional-html-scripts.hbs`)

acbeni replied:
this was done by mistake 😓 
it will be reversed

acbeni commented on app/controllers/myaccount/RecoverPasswordController.java:
```diff
@@ -51,8 +51,8 @@ public String getCmsPageKey() {
     public CompletionStage<Result> handleSuccessfulAction(final CustomerToken customerToken, final RecoverPasswordFormData formData) {
-        flash("success", "A message with further instructions has been sent to your email address");
-        return redirectToCall(recoverPasswordReverseRouter.requestRecoveryEmailPageCall());
+//        flash("success", "A message with further instructions has been sent to your email address");
```
😅  forgot to get rid of it
good catch 👍

acbeni commented:
I now added all the recovery password mecaisms in one controller `RecoverPasswordController`.

WDYT about it @lauraluiz ?

lauraluiz commented on conf/i18n/de/sunrise.yaml:
```diff
@@ -1,3 +1,5 @@
\ No newline at end of file
+  creditcard: Kreditkarte
+changePassword:
+    messageSentTitle: Eine Wiederherstellungs-E-Mail wurde gesendet, bitte überprüfen Sie Ihre E-Mail für Anweisungen zum Zurücksetzen Ihres Passworts.
```
Sorry I missed it last time, but this message should go to my-account.yaml (there's already a section for `changePassword`).
sunrise.yaml is a file that doesn't go to the Sunrise Theme, therefore it gets lost here. It's fine for the payment because it's very likely going to be removed to something more functional ;)

acbeni replied:
Sounds good to me 👍

lauraluiz commented on conf/i18n/en/sunrise.yaml:
```diff
@@ -1,3 +1,5 @@
\ No newline at end of file
+  creditcard: Credit card
+changePassword:
+  messageSentTitle: A recovery email has been sent, please check your email for instruction on how to reset your password.
```
Same here 👆

acbeni replied:
Sounds good to me 👍

lauraluiz commented on conf/routes:
```diff
@@ -100,6 +100,8 @@ GET         /:languageTag/user/logout
 GET         /:languageTag/password/recovery                                     @controllers.myaccount.RecoverPasswordController.show(languageTag: String)
 # Send recovery password email
 POST        /:languageTag/password/recovery                                     @controllers.myaccount.RecoverPasswordController.process(languageTag: String)
+#Passord recovery success page
```
*PassWord

lauraluiz commented on conf/templates/my-account-forgot-password-success.json:
```diff
@@ -0,0 +1,5 @@
+	"header" : { "$ref" : "common/header.json" },
+	"footer" : { "$ref" : "common/footer.json" },
+    "meta" : { "$ref" : "common/meta.json" }
+}
```
We don't need JSON files as input for the handlebars templates, we have the app input for that. If you keep it here just as a reminder of creating it in the Theme once you import it there, then let's remember to remove it afterwards :)

acbeni commented:
@lauraluiz i made some changes, what do you think about it now?

lauraluiz commented on common/app/com/commercetools/sunrise/framework/template/engine/AbstractHtmlContentRenderer.java:
```diff
@@ -31,7 +32,7 @@ protected AbstractHtmlContentRenderer(final UserLanguage userLanguage, final Tem
         if (cmsKey != null) {
             return cmsService.page(cmsKey, userLanguage.locales())
-                    .thenApplyAsync(cmsPage -> renderHtml(pageData, templateName, cmsPage.orElse(null)));
+                    .thenApplyAsync(cmsPage -> renderHtml(pageData, templateName, cmsPage.orElse(null)), HttpExecution.defaultContext());
```
Fix, this was missing and it shouldn't have.

lauraluiz commented on common/app/com/commercetools/sunrise/framework/viewmodels/ViewModelFactory.java:
```diff
@@ -1,5 +1,49 @@
+        return extractMessages(flash());
+    }
+
+    private static List<MessageViewModel> extractMessages(final Map<String, String> map) {
```
We can have this method protected, but is it really useful? The point of the one above is that it is part of the mechanism of saving/displaying messages, but a method that extracts messages from a map, I'm not sure where can this be applied :\
WDYT?

acbeni replied:
just to leave note, I think keeping it here is fine until the need for refactoring rise if this method proves useful 👍

lauraluiz commented on my-account/app/com/commercetools/sunrise/myaccount/addressbook/addressbookdetail/viewmodels/AddressBookPageContentFactory.java:
```diff
@@ -1,7 +1,7 @@
 
-import com.commercetools.sunrise.framework.viewmodels.content.PageContentFactory;
 import com.commercetools.sunrise.framework.viewmodels.PageTitleResolver;
+import com.commercetools.sunrise.framework.viewmodels.content.PageContentFactory;
```
From here below, everything is just formatting changes, you can ignore them :)

lauraluiz commented on conf/i18n/en/my-account.yaml:
```diff
@@ -0,0 +1,121 @@
+    bannerTitle: Women
+    bannerText: Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem
+    bottomTextTwo: Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Phasellus hendrerit. Pellentesque aliquet nibh nec urna. In nisi neque, aliquet vel, dapibus id, mattis vel, nisi. Sed pretium, ligula sollicitudin laoreet viverra, tortor libero sodales leo, eget blandit nunc tortor eu nibh. Nullam mollis. Ut justo. Suspendisse potenti. Sed egestas, ante et vulputate volutpat, eros pede ...
+  feedbackMessage: A recovery email has been sent, please check your email for instructions on how to reset your password.
```
Only this has been added.

lauraluiz commented on conf/i18n/de/my-account.yaml:
```diff
@@ -0,0 +1,119 @@
+    bannerTitle: Frauen
+    bannerText: Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem
+    bottomTextTwo: Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Phasellus hendrerit. Pellentesque aliquet nibh nec urna. In nisi neque, aliquet vel, dapibus id, mattis vel, nisi. Sed pretium, ligula sollicitudin laoreet viverra, tortor libero sodales leo, eget blandit nunc tortor eu nibh. Nullam mollis. Ut justo. Suspendisse potenti. Sed egestas, ante et vulputate volutpat, eros pede ...
+  feedbackMessage: Eine Wiederherstellungs-E-Mail wurde gesendet, bitte überprüfen Sie Ihre E-Mail für Anweisungen zum Zurücksetzen Ihres Passworts.
```
Only this has been added.

acbeni commented on common/app/com/commercetools/sunrise/framework/controllers/SunriseController.java:
```diff
@@ -22,4 +20,26 @@ public final Logger getLogger() {
+     * @param messageType type of the message to be saved (e.g. success, info)
+     * @param message the content of the message to be saved
+     */
+    protected void saveMessage(final MessageType messageType, final String message) {
```
I think that this wouldn't allow us to send multipe messages of the same type,
I'm i correct?

lauraluiz replied:
Yep, you are correct. That's a limitation of Play itself because flash doesn't allow to save objects (as it's a cookie it's also not recommended). But there are workarounds for that:
1) set in flash your own message type and extract it into view models (just around 4 lines of simple code)
2) have into the same message type multiple messages (the approach might change depending on the situation)
3) we can also easily extend the system by using the message types as prefixes and allow multiple messages without breaking changes

(Just a summary of what we talked about to have it written 😉 )

acbeni commented:
@lauraluiz it looks good to me 👍  now its much better and the messages can be used for more general purposes 🥇 
If you think its good enough you can approve my changes ( 😅  since i am te one who started the PR i can't accept your changes) as of then i can update the theme to support the new feature.

acbeni commented:
@lauraluiz i updated the theme version, and deleted the temporary files, tested sunrise locally and it all seems good 👍 
guess we'll merge the changes.

lauraluiz commented:
Yep, it does look good! You can merge changes if you want :)
